app/views/courses/TeacherClassView.coffee

```diff
@@ -0,0 +1,176 @@
+    for student in @students.models
+      # TODO: this is a weird hack
+      studentsStub = new Users([ student ])
+      student.latestCompleteLevel = helper.calculateLatestComplete(@classroom, @courses, @campaigns, @courseInstances, studentsStub)
+      
+    classroomsStub = new Classrooms([ @classroom ])
+    @progressData = helper.calculateAllProgress(classroomsStub, @courses, @campaigns, @courseInstances, @students)
+    super()
+    
+  copyCode: ->
+    @$('#join-code-input').val(@classCode).select()
+    @tryCopy()
+  
+  copyURL: ->
+    @$('#join-url-input').val(@joinURL).select()
+    @tryCopy()
+    
+  tryCopy: ->
+    try
+      document.execCommand('copy')
+      application.tracker?.trackEvent 'Classroom copy URL', category: 'Courses', classroomID: @classroom.id, url: @joinURL
+    catch err
+      console.log('Oops, unable to copy', err)
+
+  onClickAddStudents: (e) =>
+    modal = new InviteToClassroomModal({ classroom: @classroom })
+    @openModalView(modal)
+    @listenToOnce modal, 'hide', @render
+    
+  onClickEnrollStudent: (e) ->
+    # TODO
+  
+  sortByName: (e) =>
+    if @sortValue == 'name'
+      @sortDirection = -@sortDirection
+    else
+      @sortValue = 'name'
+      @sortDirection = 1
+      
+    dir = @sortDirection
+    @students.comparator = (student1, student2) ->
+      return (if student1.get('name') < student2.get('name') then -dir else dir)
+    @students.sort()
+    
+  sortByProgress: (e) =>
+    if @sortValue == 'progress'
+      @sortDirection = -@sortDirection
+    else
+      @sortValue = 'progress'
+      @sortDirection = 1
+      
+    dir = @sortDirection
+    @students.comparator = (student1, student2) ->
+      l1 = student1.latestCompleteLevel
+      l2 = student2.latestCompleteLevel
+      if l1.courseNumber < l2.courseNumber
+        return -dir
+      else if l1.levelNumber < l2.levelNumber
+        return -dir
+      else
+        return dir
+    @students.sort()
+  
+  getSelectedStudentIDs: ->
+    $('.student-row .checkbox-flat input:checked').map (index, checkbox) ->
+      $(checkbox).data('student-id')
+    
+  ensureInstance: (courseID) ->
+    
+  onClickEnroll: (e) ->
+    userID = $(e.target).data('user-id')
+    user = @students.get(userID)
+    modal = new ActivateLicensesModal { @classroom, user, users: @students }
+    @openModalView(modal)
+    modal.once 'redeem-users', -> document.location.reload()
+    application.tracker?.trackEvent 'Classroom started enroll students', category: 'Courses'
+    
+  onClickBulkAssign: ->
+    courseID = $('.bulk-course-select').val()
+    courseInstance = @courseInstances.getByCourseAndClassroom(courseID, @classroom)
+    members = @getSelectedStudentIDs().toArray()
+    if courseInstance
+      # TODO: Only assign the selected ones
+      for studentID in members
+        courseInstance.addMember studentID, {
+          success: =>
```

Comment:
This is going to send off a series of network requests which will lead to a race condition which will probably result in only some of the classroom members appearing on the client. Possible solutions:

1. Add serially.
1. Create a new endpoint which adds multiple members at once.
1. Wait until all responses finish and then reload the page.